straxen/plugins/peaks/peak_subtyping.py

```diff
@@ -0,0 +1,621 @@
+        self.large_s2_threshold += np.sqrt(self.large_s2_threshold_ne) * self.se_span
+        self.mis_s2_threshold = self.mis_s2_fac * self.se_gain
+        # we need to set those window sizes as integers because they are used to define containers
+        self.s1_s2_window = int(self.s1_s2_window_fac * self.drift_time_max)
+        self.after_s1_window_ext = int(self.after_s1_window_ext_fac * self.drift_time_max)
+        self.after_s2_window_ext = int(self.after_s2_window_ext_fac * self.drift_time_max)
+
+    def get_window_size(self):
+        return 10 * max(self.s1_s2_window, self.after_s1_window_ext, self.after_s2_window_ext)
+
+    @staticmethod
+    def mark_peaks_after_s1s(
+        peaks, subtype_dtype,
+        mis_s2_threshold, large_s2_threshold, other_large_s2_fac,
+        s1_s2_window, after_s1_window_ext, after_s2_window_ext):
+        '''
+        Look after each S1s and classify:
+        1. mark the largest peak within s1_s2_window as pS2 (exceeding large_s2_threshold)
+        2. extend window after pS2
+        3. mark all other large S2s as olS2. 
+        4. mark all other small S2s as photoionization (PH)
+
+        if pS2 identified, extend the window to after_s2_window_ext after such pS2 and mark S2s:
+        1. other large S2 if an S2 is larger than large_s2_threshold and half the pS2 size
+        2. photoionization if not olS2
+
+        The order of subtyping assignment is important
+        and reveals the priorities of the subtypings
+        '''
+
+        # prepare for referred variables
+        mask = np.ones(len(peaks), dtype=subtype_dtype) * PeakSubtyping.Undefined
+        type_1_mask = (peaks['type'] == 1)
+        type_2_mask = (peaks['type'] == 2)
+
+        _s1_area = peaks['area'][type_1_mask]
+        _s2_area = peaks['area'][type_2_mask]
+
+        base_times = peaks['center_time']
+        base_areas = peaks['area']
+        base_indices = np.arange(len(peaks))
+
+        # find S1-pS2 pairing
+        _peaks = peaks[type_2_mask].copy()
+        # Only use center_time because we used center_time to define the drift time
+        _peaks['time'] = _peaks['center_time']
+        # touchng_windows needs non-zero length of things
+        _peaks['endtime'] = _peaks['time'] + 1
+
+        _window = np.zeros(type_1_mask.sum(), dtype=strax.time_fields)
+        _window['time'] = peaks['center_time'][type_1_mask]
+        _window['endtime'] = _window['time'] + s1_s2_window
+        # Here the _window can be overlapping
+        tw_s1_ps2 = strax.touching_windows(_peaks, _window)
+
+        # Find index of max S2 of each group of S2s following S1
+        _ps2_max_indices = np.ones(len(_window), dtype=np.int32) * -1
+        for i, tw_12 in enumerate(tw_s1_ps2):
+            if tw_12[1] - tw_12[0] == 0:
+                continue
+            else:
+                # do we have pS2 or a bunch of small S2s
+                _ps2_max_indices[i] = np.argmax(_s2_area[tw_12[0]:tw_12[1]]) + tw_12[0]
```

Comment:
Here too, I would do something like:

```python
index_s2_with_max_area = ...
 _ps2_max_indices[i] = index_s2_with_max_area

Reply:
For readability, I agree. But will this slow down the calculation?

Reply:
I never tested it, but I do not think so. At least I doubt that it will be noticeable